design/ratelimit-design.md

```diff
@@ -1,175 +1,501 @@
+A CRD could be used -- either `HTTPProxy` or a stand-alone one -- to define actual rate limits, and a controller could convert the CRDs into configuration in the Lyft RLS format.
+This option was discarded (for now) because we have potential users who are interested in using other rate limiting services, so just providing an integration with the Lyft implementation is not sufficient. 
+Future work could still be done to enable rate limits to be defined via CRD (as part of `HTTPProxy`, or stand-alone) and automatically configured with 1+ underlying RLS implementations.
+
+### Contour as external RLS
+Contour could function as an external RLS itself.
+This would put Contour in the data path for requests.
+It would simplify deployment and configuration for the user, at a cost of significant additional complexity for Contour.
+
+### Unique RLS per TLS virtual host
+It may be desirable for each virtual host to be able to configure a different RLS.
+This is not currently possible for non-TLS virtual hosts, because they all share a single HTTP Connection Manager (HCM)/rate limit filter config.
+However, this could be supported for TLS virtual hosts because they each have their own HCM and rate limit filter config.
+To support this, the same `rateLimitServer` struct used in the config file could be added as a field to `HTTPProxy.Spec.VirtualHost`:
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: proxy
+  namespace: projectcontour
+spec:
+  virtualhost:
+    fqdn: local.projectcontour.io
+    tls:
+      secretName: local-tls
+    # rateLimitServer optionally defines a non-default RLS to use
+    # for this HTTPProxy. This field can only be specified for root
+    # TLS-enabled HTTPProxies.
+    rateLimitServer:
+      # extensionRef is an object reference to the RLS ExtensionService.
+      extensionRef:
+        name: ratelimit
+      # domain is passed to the RLS for all rate limit requests.
+      # Defaults to "contour".
+      domain: contour
+      # failOpen defines whether to fail open or closed. If false, then if the RLS
+      # cannot be reached or does not return a valid rate limit decision within the
+      # specified timeout, the client will receive a 429 response for their request.
+      #
+      # This sets the "denyOnFailure" field in the Envoy config.
+      failOpen: true
+...
 ```
-domain: <unique domain ID>
-descriptors:
-  - key: <rule key: required>
-    value: <rule value: optional>
-    rate_limit: (optional block)
-      unit: <see below: required>
-      requests_per_unit: <see below: required>
-    descriptors: (optional block)
-      - ... (nested repetition of above)
-```
 
-#### Rate limit definition
+The `HTTPProxy` processor would ensure that the `rateLimitServer` field could only be specified for root TLS-enabled `HTTPProxies`.
+
+This alternative may be pursued in the future, as we gather more information on how users want to configure this feature.
```

Comment:
Some reasons why we went against this could be that we don't anticipate (until users tell us so) that there will be the kind of many-to-many virtualhost/route to ratelimit server relationship that you might get with other extension services and that starting with this structure gives us the ability to be flexible without adding too much complexity up front.